chain/src/txhashset/txhashset.rs

```diff
@@ -550,9 +550,10 @@ where
 	let res: Result<T, Error>;
 	let rollback: bool;
 
-	// We want to use the current head of the most work chain unless
-	// we explicitly rewind the extension.
-	let head = batch.head()?;
+	// We want to use the current head of header chain here.
+	// Caller is responsible for rewinding the header MMR back
+	// to a previous header as necessary when processing a fork.
+	let head = batch.header_head()?;
```

Comment:
👍 the puzzle of `header head` and `body head` :-) thanks very much for finding this long standing bug.

I'm a little bit confused on the comment:
 > Caller is responsible for rewinding the header MMR back to a previous header as necessary when processing a fork.

Does it ONLY refer to a calling of the `extension.force_rollback()` to set that flag?

Reply:
Say we are at block height 5 on both main chain and header chain.

```
1 <- 2 <- 3 <- 4 <- 5
```

If we now receive a block at height 4' that forks from the main chain then we need to "rewind" back to block height 3 to be able to process this new block.

```
1 <- 2 <- 3 <- 4'
```

We would then rollback the MMR as this block did not increase the total work, i.e. its not a reorg, its just a fork.

If we then saw 5' that _did_ increase total work compared to 5 then we would rewind as before and then persist the MMR and update the `head` (and `header_head`) to reflect the chain after the reorg.